apps/labs/posts/unlocking-c-level-performance-in-df-apply.md

```diff
@@ -0,0 +1,466 @@
+        <td>0.378354</td>
+        <td>-0.501367</td>
+      </tr>
+      <tr>
+        <th>9998</th>
+        <td>0.495766</td>
+        <td>1.713530</td>
+        <td>-1.892357</td>
+      </tr>
+      <tr>
+        <th>9999</th>
+        <td>0.644388</td>
+        <td>-1.619920</td>
+        <td>-0.228884</td>
+      </tr>
+    </tbody>
+  </table>
+  <p>10000 rows × 3 columns</p>
+</div>
+
+#### Measuring compilation and boxing/unboxing overhead
+
+First, to give an idea of how much overhead compilation and boing/unboxing creates,
+we will use a simple function that returns the input Series without modification.
+
+```python
+f = lambda x: x
+```
+
+```python
+# Timing the compilation time
+# To estimate this, we will use the numba engine on a very small DataFrames
+small_df = pd.DataFrame({"a": [1]})
+%time small_df.apply(f, engine="numba", axis=1)
+```
+
+    CPU times: user 312 ms, sys: 6.93 ms, total: 319 ms
+    Wall time: 321 ms
+
+<div>
+<table border="1" class="dataframe">
+  <thead>
+    <tr>
+      <th></th>
+      <th>a</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <th>0</th>
+      <td>1</td>
+    </tr>
+  </tbody>
+</table>
+</div>
+
+```python
+# Notice how the compilation overhead disappears on the second run
+# because of the caching of the compiled function
+%time small_df.apply(f, engine="numba", axis=1)
+```
+
+    CPU times: user 985 µs, sys: 15 µs, total: 1e+03 µs
+    Wall time: 1e+03 µs
+
+<div>
+<table border="1" class="dataframe">
+  <thead>
+    <tr>
+      <th></th>
+      <th>a</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <th>0</th>
+      <td>1</td>
+    </tr>
+  </tbody>
+</table>
+</div>
+
+Now, let's measure the overhead of the boxing/unboxing of the input DataFrame.
+To do this, we will compare the speed of apply with numba with raw=True and raw=False.
+
+Although doing a raw apply on the DataFrame still does have some overhead (mainly due to converting the DataFrame into a 2D ndarray),
+it is minimal compared to the total execution time.
+
+```python
+%timeit df.apply(f, engine="numba", axis=1)
+```
+
+    330 ms ± 24.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
+
+```python
+%timeit df.apply(f, engine="numba", axis=1, raw=True)
+```
+
+    307 µs ± 44.3 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
+
+We can see here that the difference is several orders of magnitude.
+This is because the unboxing process currently individually unboxes each resultant Series to a Python object
+before concatenating them together to build the result DataFrame.
+
+In future pandas releases, this can be optimized by concatenating the results of apply inside of the numba
+engine (which would only require one final unobxing call for the concatenated DataFrame),
+which should bring the speed of apply on Series in numba on par with the speed of
+apply on the raw numpy arrays.
+
+Despite this, the numba engine is still much faster than the Python engine, as seen below.
```

Comment:
My impression here is that it helps to spell it out and not let the reader do the math. ~6x speedup for `Series`, and ~6000x for raw numpy arrays!

Reply:
I'll add a comment to the end, since that also talks about the parallel speed.

This benchmark in particular has also gotten stale, somehow the Python engine for apply caught up (maybe a regression got fixed somewhere), so the numba impl is now a bit slower than Python in the noop case. I'll take a look to see if we can reduce overhead a little more.